backend/editor/controllers/search_controller.py

```diff
@@ -0,0 +1,374 @@
+        language_code, search_value = search_value.split(":", maxsplit=1)
+        language_code = language_code.lower()
+
+    def get_search_query() -> str | None:
+        if search_value.startswith('"') and search_value.endswith('"'):
+            return search_value if len(search_value) > 2 else None
+
+        if language_code is not None:
+            normalized_text = parser_utils.normalize_text(search_value, language_code)
+        else:
+            normalized_text = parser_utils.normalize_text(search_value)
+
+        # If normalized text is empty, no searches are found
+        if normalized_text.strip() == "":
+            return None
+
+        tokens = normalized_text.split("-")
+
+        def get_token_query(token: str) -> str:
+            token = "+" + token
+            if len(token) > 10:
+                return token + "~2"
+            elif len(token) > 4:
+                return token + "~1"
+            else:
+                return token
+
+        return "(" + " ".join(map(get_token_query, tokens)) + ")"
+
+    search_query = get_search_query()
+
+    if search_query is None:
+        return None
+
+    if language_code is not None:
+        search_query = f"tags_ids_{language_code}:{search_query}"
+
+    return search_query
+
+
+def build_cypher_name_search_term(project_id: str) -> tuple[str, str, str]:
+    query = f"""
+            CALL db.index.fulltext.queryNodes("{project_id}_SearchTagsIds", $search_query)
+            YIELD node, score
+            WHERE score > 0.1
+            WITH node.id AS nodeId
+            WITH COLLECT(nodeId) AS nodeIds
+        """
+
+    where_clause = "n.id IN nodeIds"
+
+    order_clause = "WITH n, apoc.coll.indexOf(nodeIds, n.id) AS index ORDER BY index"
+
+    return query, where_clause, order_clause
+
+
+def build_cypher_query(
+    query: Query, skip: int, limit: int
+) -> tuple[str, str, dict[str, str]] | None:
+    lucene_name_search_queries = list(
+        filter(
+            lambda q: q is not None, map(build_lucene_name_search_query, query.name_search_terms)
+        )
+    )
+
+    cypher_filter_search_terms = [
+        term.build_cypher_query(index) for index, term in enumerate(query.filter_search_terms)
+    ]
+
+    full_text_search_query, order_clause = "", ""
+    query_params = {}
+
+    if lucene_name_search_queries:
+        (
+            full_text_search_query,
+            name_filter_search_term,
+            order_clause,
+        ) = build_cypher_name_search_term(query.project_id)
+        query_params["search_query"] = " AND ".join(lucene_name_search_queries)
+        cypher_filter_search_terms.append((name_filter_search_term, None))
+
+    query_params |= {
+        get_query_param_name(index): cypher_filter_search_term[1]
+        for index, cypher_filter_search_term in enumerate(cypher_filter_search_terms)
+    }
+
+    combined_filter_query = (
+        f"WHERE {' AND '.join(map(lambda x: x[0], cypher_filter_search_terms))}"
```

Comment:
I python culture we prefer list comprehension (more readable IMO):

```suggestion
        f"WHERE {' AND '.join(x[0] for x in cypher_filter_search_terms)}"
```